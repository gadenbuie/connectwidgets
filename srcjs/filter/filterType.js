/**
 * Filter types
 */
 export const FilterType = {
  owner: 'owner',
  type: 'type',
  tag: 'tag',
};
