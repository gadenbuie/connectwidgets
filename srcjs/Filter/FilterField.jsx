import PropTypes from 'prop-types';
import { debounce } from '@/utils';
import FilterOptions from './FilterOptions';
import FiltersSelected from './FiltersSelected';
import { FilterType } from './filterType';
import { FilterContext } from './stateUtils';
import './filterField.scss';

const fieldClass = {
  base: 'rscfilter-field__input-wrap',
  dropdown: 'rscfilter-field__input-wrap rscfilter-field__input-wrap--dropdown',
};

function FilterField({ label, placeholder, type }) {
  const inputRef = React.createRef();

  const context = React.useContext(FilterContext);
  const isDropdown = type === FilterType.type;

  const prefilteringHandler = ev => {
    const prefix = ev.target.value.trim();
    context.searchFieldOptions(type, prefix);
  }

  const handleEsc = ev => {
    if (ev.key === "Escape") {
      ev.stopPropagation();
      inputRef.current.blur();
    }
  }

  return (
    <div className="rscfilter-field">
      <div className={isDropdown ? fieldClass.dropdown : fieldClass.base}>
        <label className="rscfilter-field__name">{label}</label>
        <input
          className="rscfilter-field__input"
          ref={inputRef}
          name={type}
          placeholder={placeholder}
          onChange={debounce(300, prefilteringHandler)}
          onKeyUp={handleEsc}
        />
        <FilterOptions type={type} />
      </div>
      <FiltersSelected type={type} />
    </div>
  );
}

FilterField.propTypes = {
  type: PropTypes.string,
  label: PropTypes.string,
  placeholder: PropTypes.string,
}

export default FilterField;
