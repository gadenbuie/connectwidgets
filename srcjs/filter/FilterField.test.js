import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import FilterField from './FilterField';
import { FilterType } from './filterType';

jest.useFakeTimers();

const fulfillDebounce = () => jest.runAllTimers();
const inputEvWith = value => ({ target: { value }});
const fakeContext = {
  options: { [FilterType.owner]: [{
      name: 'admin',
      label: 'admin',
      isSelected: false,
    }]
  },
  searchFieldOptions: jest.fn(),
  updateSelection: jest.fn(),
};
jest.spyOn(React, 'useContext').mockReturnValue(fakeContext);

const mkWrapper = () => shallow(<FilterField label="Owner" placeholder="Filter people" type={FilterType.owner} />);

describe('<FilterField />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('renders properly', () => {
    const wrapper = mkWrapper();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('search field options on change and after debounce', () => {
    const wrapper = mkWrapper();

    wrapper.find('input').simulate('change', inputEvWith('tommy'));
    expect(fakeContext.searchFieldOptions).not.toHaveBeenCalled();

    fulfillDebounce();
    expect(fakeContext.searchFieldOptions).toHaveBeenCalledWith(FilterType.owner, 'tommy');
  });
});
