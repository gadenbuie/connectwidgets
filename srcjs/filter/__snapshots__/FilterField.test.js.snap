// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterField /> renders properly 1`] = `
<div
  className="rscfilter-field"
>
  <div
    className="rscfilter-field__input-wrap"
  >
    <label
      className="rscfilter-field__name"
    >
      Owner
    </label>
    <input
      className="rscfilter-field__input"
      name="owner"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="Filter people"
    />
    <FilterOptions
      type="owner"
    />
  </div>
  <FiltersSelected
    type="owner"
  />
</div>
`;
